### PR TITLE
Implement Array/TypedArray find index

### PR DESCRIPTION
**Pull request by Yuyupo**
JerryScript-DCO-1.0-Signed-off-by: Daniella Barsony bella@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c [hunk 1]
```diff
@@ -1501,19 +1501,18 @@ ecma_builtin_typedarray_prototype_sort (ecma_value_t this_arg, /**< this argumen
...
  * @return ecma value
  *         Returned value must be freed with ecma_free_value.
  */
 static ecma_value_t
-ecma_builtin_typedarray_prototype_find (ecma_value_t this_arg, /**< this argument */
-                                        ecma_value_t predicate, /**< callback function */
-                                        ecma_value_t predicate_this_arg) /**< this argument for
-                                                                          *   invoke predicate */
+ecma_builtin_typedarray_prototype_find_helper (ecma_value_t this_arg, /**< this argument */
+                                              ecma_value_t predicate, /**< callback function */
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: These arguments misaligned by 1 space.

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c [hunk 2]
```diff
@@ -1559,15 +1558,45 @@ ecma_builtin_typedarray_prototype_find (ecma_value_t this_arg, /**< this argumen
...
   }
 
   return ECMA_VALUE_UNDEFINED;
+} /* ecma_builtin_typedarray_prototype_find_helper */
+
+/**
+ * The %TypedArray%.prototype object's 'find' routine
+ *
+ * See also:
+ *          ECMA-262 v6, 22.2.3.10
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: Missing `@return` description.

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c [hunk 3]
```diff
@@ -1559,15 +1558,45 @@ ecma_builtin_typedarray_prototype_find (ecma_value_t this_arg, /**< this argumen
...
+{
+  return ecma_builtin_typedarray_prototype_find_helper (this_arg, predicate, predicate_this_arg, true);
 } /* ecma_builtin_typedarray_prototype_find */
 
+/**
+ * The %TypedArray%.prototype object's 'findIndex' routine
+ *
+ * See also:
+ *          ECMA-262 v6, 22.2.3.11
+ */
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: Ditto.

**Comment by Yuyupo** [comment]
@rerobika Fixed the style errors you mentioned.

File: tests/jerry/es2015/array-prototype-find-index.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+function func (element) {
+  return element > 8;
+}
+
+/* ES v6.0 22.1.3.8.8.c
+   Checking behavior when the first element deletes the second */
+function f() { delete arr[1]; };
+var arr = [0, 1, 2, 3];
+Object.defineProperty(arr, '0', { 'get' : f });
+Array.prototype.findIndex.call(arr, func);
```
Comment (galpeter) [inline tests/jerry/es2015/array-prototype-find-index.js]: What is the expected here? Should this crash? Should return nothing?

File: tests/jerry/es2015/typedArray-find-index.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    return element > 1;
+  }
+
+  assert (src_array.findIndex (isPrime) === undefined);
+
+  src_array = new TypedArray ([4, 5, 8, 12]);
+  array_index = 0;
+  assert (src_array.findIndex (isPrime) === 1);
+
+  // Checking behavior when the given object is not %TypedArray%
```
Comment (galpeter) [inline tests/jerry/es2015/typedArray-find-index.js]: Such test case should also be present in the Array test file above.

File: tests/jerry/es2015/array-prototype-find-index.js (hunk 4 above)
Comment (Yuyupo) [reply tests/jerry/es2015/array-prototype-find-index.js]: All it does is iterate through an array which has a size of "n", so we expect it to have n items but while iterating through it an item gets deleted and the size stays the same. => hence ecma_op_object_find throws an error and we have more branch coverage

**Comment by Yuyupo** [comment]
@galpeter Added the missing test case

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c [hunk 6]
```diff
@@ -1559,15 +1558,51 @@ ecma_builtin_typedarray_prototype_find (ecma_value_t this_arg, /**< this argumen
...
     {
-      return element_value;
+      return is_find ? element_value : ecma_make_uint32_value (buffer_index);
     }
-
+    buffer_index++;
     ecma_free_value (element_value);
   }
 
   return ECMA_VALUE_UNDEFINED;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: `22.1.3.9. 9` Return -1. FindIndex should return -1 if no matching element is found.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 7]
```diff
@@ -2043,7 +2047,7 @@ ecma_builtin_array_prototype_object_find (ecma_value_t predicate, /**< callback
       if (ecma_op_to_boolean (call_value))
       {
         /* 8.f */
-        ret_value = ecma_copy_value (get_value);
+        ret_value = is_find ? ecma_copy_value (get_value) : current_index;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: On line 2064, the return value must be `-1` for findIndex as well.

Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Also please add a test case for it.

Comment (Yuyupo) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Added the `-1` return value to line 2064 and modified the tests accordingly. :)

**Pull request merged**